src/lib/storage/segment_access_statistics.hpp

```diff
@@ -0,0 +1,379 @@
+#pragma once
+
+#include <array>
+#include <atomic>
+#include <chrono>
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <string>
+
+#include <types.hpp>
+#include <storage/pos_list.hpp>
+
+namespace opossum {
+class Table;
+// -----------------------------------------------------------------------------------------------------------------
+enum class SegmentAccessType {
+  Other,
+  IteratorCreate,
+  IteratorSeqAccess,
+  IteratorIncreasingAccess,
+  IteratorRandomAccess,
+  IteratorDereference,
+  AccessorCreate,
+  AccessorAccess,
+  DictionaryAccess,
+  // number of elements in enum
+    Count
+};
+
+// -----------------------------------------------------------------------------------------------------------------
+enum SegmentType {
+  Dictionary,
+  FrameOfReference,
+  LZ4,
+  Reference,
+  RunLength,
+  ValueS
+};
+
+// -----------------------------------------------------------------------------------------------------------------
+
+class SegmentAccessTypeTools {
+ public:
+  inline static const std::string headers = "Other,IteratorCreate,IteratorSeqAccess,IteratorIncreasingAccess,"
+                                            "IteratorRandomAccess,IteratorDereference,AccessorCreate,AccessorAccess,"
+                                            "DictionaryAccess";
+
+  static SegmentAccessType iterator_access_pattern(const std::shared_ptr<const PosList>& positions);
+
+ private:
+  // There are five states:
+  // 0 (unknown), equivalent to IteratorSeqAccess
+  // 1 (sequentially increasing), difference between two neighboring elements is 0 or 1.
+  // 2 (increasing randomly)
+  // 3 (sequentially decreasing), difference between two neighboring elements is -1 or 0.
+  // 4 (decreasing randomly)
```

Comment:
Have you seen these?

Reply:
Query 7, for example, iterates through data sequentially [increasing|decreasing], randomly [increasing|decreasing] and randomly.
Idk if state 3 or 4 are ever reached.